feat(frontend): Feature reduce pipeline graph. Fixes #4924

StefanoFioravanzo opened the pull request:
**Description of your changes:**

Add a "Transitive Reduction" mechanism to simplify the graph in the Pipeline Details and in the Run Details pages. Activating this function will remove any redundant dependencies in the graph, thus making it simpler to read in many instances.

In case the graph cannot be reduced, the toggle will be disabled. Toggling again this option will revert the graph back to its original layout, so the user can switch it on and off to compare the two views, if needed.

More details at #4924 

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
<!--
   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.
-->
- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).
<!--
    **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update before release.
-->

Udiknedormin commented:
Isn't something similar implemented on `PipelineVolume`'s `after` method? Wouldn't it be better to reduce dependencies directly in the produced Argo yaml instead of the UI?

Bobgy commented:
Thank you @StefanoFioravanzo!
Love this idea!

I prefer the generic graph simplification at UI layer.

StefanoFioravanzo commented:
@Udiknedormin I answered to your question in the issue. Thanks @Bobgy !

Bobgy commented on frontend/src/lib/StaticGraphParser.ts:
```diff
@@ -263,3 +263,51 @@ export function createGraph(workflow: Workflow): dagre.graphlib.Graph {
+  return graph;
+}
+
+export function compareGraphEdges(graph1: dagre.graphlib.Graph, graph2: dagre.graphlib.Graph) {
```
I'm not seeing why this is useful.
I think hiding the sth dynamically in the UI is generally a bad design, users will be confused, because when the switch doesn't show up. There's no clue why it doesn't.

In this specific case, I think the simplest design is enough:
* always show the simplify graph switch

StefanoFioravanzo replied:
So this is used both in RunDetails and PipelineDetails pages to compare the reduced graph vs the original one. If there is a difference then the reduced one gets picked to be displayed.

Bobgy commented on frontend/src/pages/PipelineDetails.tsx:
```diff
@@ -288,14 +300,26 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
                     />
 
+                    <div className={commonCss.transitiveReductionSwitch}>
+                      <SwitchWithLabel
```
I think it'll be better if we have a tooltip on the switch/ or next to the switch that explains how we are simplifying the graph. A short paragraph, probably with a link to more details is great.

Because, the simplified dependencies might mean passing parameters/artifacts between steps and without explaining how we are simplifying the graph, it's will be confusing why those dependencies do not show up.

Bobgy replied:
For tooltip, you may use https://github.com/kubeflow/pipelines/blob/768317aee33d1c00a46637549fc2753aa357e0ab/frontend/src/atoms/HelpButton.tsx#L23 or https://github.com/kubeflow/pipelines/blob/768317aee33d1c00a46637549fc2753aa357e0ab/frontend/src/components/CustomTable.tsx#L347

StefanoFioravanzo replied:
Added a tooltip as part of the refactoring into a component 6d4c769c

StefanoFioravanzo replied:
I used the same tooltip as in `HelpButton`, so this code pattern is now repeated in two places. I tried to implement a new atom `CardTooltip` but I had some issues with passing `props.children`. This is what I tried to implement:

```jsx
const NostyleTooltip = withStyles({
  tooltip: {
    backgroundColor: 'transparent',
    border: '0 none',
    color: color.secondaryText,
    fontSize: fontsize.base,
    maxWidth: 220,
  },
})(Tooltip);

interface CardTooltipProps {
  helpText?: ReactNode;
}
export const CardTooltip: React.FC<CardTooltipProps> = props => {
  return (
    <NostyleTooltip
      title={
        <Card>
          <CardContent>{props.helpText}</CardContent>
        </Card>
      }
      interactive={true}
      leaveDelay={400}
      placement='top'
    >
      {props.children}
    </NostyleTooltip>
  );
};
```

but I get an error in `{props.children}`. Apparently the `NostyleTooltip` cannot accept a `ReactNode`: `Type 'ReactNode' is not assignable to type 'ReactElement<any, string | ((props: any) => .......`. This is a bit beyond me, do you have any tip on how to fix it or debug further?

StefanoFioravanzo replied:
@Bobgy I gave this a second look and I figured out what was wrong, see my latest commit: https://github.com/kubeflow/pipelines/pull/4925/commits/80f3f1dcdc33631f8476ffec289671e218905e17

Bobgy reviewed (commented):
Read through code in details, left some thoughts.

A general question: why did you decide to show simplified graph by default?
I think in smaller graphs, seeing all the dependencies can be useful because each dependency means some parameter or artifact passing happens between the two tasks.

In bigger graphs, we might let users click the simplify graph switch by themselves.

Bobgy commented on frontend/src/atoms/SwitchWithLabel.tsx:
```diff
@@ -0,0 +1,43 @@
+  width?: number;
+}
+
+const SwitchWithLabel = (props: SwitchWithLabelProps) => {
```
considering the request to add tooltip on top of it, I think it's better building a component under `frontend/src/components/` that's specific to this use-case

StefanoFioravanzo replied:
I created a new component and removed the atom, check out 6d4c769c

StefanoFioravanzo commented:
@Bobgy Thank you for the detailed review! I will go through all of your comments in a couple of weeks, after the Christmas break.

StefanoFioravanzo commented on frontend/src/lib/StaticGraphParser.test.ts:
```diff
@@ -780,4 +787,52 @@ describe('StaticGraphParser', () => {
+    it('reduces redundant workflow', () => {
+      const g = createGraph(newRedundantWorkflow());
+      const reduced = transitiveReduction(g);
+      expect(reduced.nodeCount()).toEqual(3);
+      expect(reduced.edgeCount()).toEqual(2);
```
See here 1c7906a1

StefanoFioravanzo commented on frontend/src/lib/StaticGraphParser.ts:
```diff
@@ -263,3 +263,51 @@ export function createGraph(workflow: Workflow): dagre.graphlib.Graph {
+ *
+ * @param graph The dagre graph object
+ */
+export function transitiveReduction(graph: dagre.graphlib.Graph) {
```
See here for deep copy improvement: beb26374
Time complexity comment: aa4d3403
Big graphs safeguard: 8f53e485  -> This change actually introduces an issue with the tests. I will explain more in a comment below

StefanoFioravanzo commented on frontend/src/pages/PipelineDetails.tsx:
```diff
@@ -198,9 +206,13 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
         !window['safari'] || (typeof 'safari' !== 'undefined' && window['safari'].pushNotification),
       );
 
+    const _graph =
```
Oh I didn't know this, I will take a better look at the style guide. Fixed here d6b34e88

StefanoFioravanzo commented:
@Bobgy I think I addressed all your comments. As for this:

> why did you decide to show simplified graph by default?

I agree that in general we should leave it off as default, I changed that here 657bd6db

Also, following up on this https://github.com/kubeflow/pipelines/pull/4925#discussion_r551419777 :
Returning `undefined` in `transitiveReduction` makes typescript complain in the tests here https://github.com/kubeflow/pipelines/blob/208618bcd8f6182bd13e13f7b75cd6164454ffae/frontend/src/lib/StaticGraphParser.test.ts#L828 saying that `reduced` ` Object is possibly 'undefined'.`. That's obviously true, and I am not sure if we had better just return the unchanged graph instead of `undefined` or if we should handle this situation in a different way

Bobgy commented:
Thanks, I'll take further look this week

StefanoFioravanzo commented:
Hey @Bobgy any update here? 😃

Bobgy commented on frontend/src/pages/PipelineDetails.tsx:
```diff
@@ -521,8 +546,14 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
+    const graph = await this._createGraph(templateString);
+    let reducedGraph = graph ? transitiveReduction(graph) : null;
+    if (graph && reducedGraph && compareGraphEdges(graph, reducedGraph)) {
+      reducedGraph = null; // disable reduction switch
```
Personal opinion:
I think there's no need to disable the switch, when reduced graph is the same.
It's understandable -- when the graph is already simple, reducing it has no effect. We've got tooltip on the 
switch too to explain this.

What do you think? anyway, this is not a blocker

StefanoFioravanzo replied:
I would leave it disabled so that it's obvious that this features is not available for that particular pipeline. Otherwise users might try clicking it anyway, see that nothing changes, and wonder what is happening. I wouldn't rely on people actually reading and understanding what's in the tooltip, we can just prevent any potential misunderstanding by disabling the switch

Bobgy commented on frontend/src/pages/PipelineDetails.tsx:
```diff
@@ -521,8 +546,14 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
     this.props.updateToolbar({ breadcrumbs, actions: toolbarActions, pageTitle });
 
+    const graph = await this._createGraph(templateString);
+    let reducedGraph = graph ? transitiveReduction(graph) : null;
```
nit: for consistency, always use `undefined` when possible.

Bobgy reviewed (commented):
@StefanoFioravanzo Sorry for the late reply.

Code mostly looks good, I've read through it in details.

Despite some minor nit pickings -- can you write unit tests in `RunDetails.test.tsx` and `PipelineDetails.test.tsx` to verify UI behavior after clicking on the simplify graph switch?

StefanoFioravanzo commented:
@Bobgy about unit tests, I have close to zero experience in writing tests for React components. So please bear with me a little bit here :) See the last commit https://github.com/kubeflow/pipelines/pull/4925/commits/d080cc94c70bc8904d76fe3fa56f19883d38fbd6 where I try to test a diamond-shaped pipeline. Creating the graph and testing for its nodes and edges works. Now I want to query for the switch component, simulate its click and test that the graph is missing the edge "node1 to node4". How can I query for the switch element and click it?

Bobgy commented:
@StefanoFioravanzo that's a good start,
The test is using @testing-library/react, so you can find documentation in https://testing-library.com/docs/queries/about (note that our version looks slightly outdated).

There's a section that introduces all types of query you can use,
for this case
```
getByText: Not useful for forms, but this is the number 1 method a user finds most non-interactive elements (like divs and spans).
getByLabelText: Only really good for form fields, but this is the number one method a user finds those elements, so it should be your top preference.
```
either of this sounds reasonable, because users would find your switch looking at the text on that switch.

You can use https://testing-library.com/docs/dom-testing-library/api-events library to fire the click event easily. (I believe we didn't install it, but we've decided to use @testing-library/ family, so you can just install it).


===

Here are the information,

Anyway, if you still cannot figure out in reasonable time. It's OK to log a remaining issue and leave it there. This PR is already great work!

/cc @zijianjoy to also do some review

StefanoFioravanzo commented:
@Bobgy Thanks that suggestion was spot on! I updated the commit with a complete unit test for the RunDetails page. I gave a look at PipelineDetails as well, but there doesn't seem to be an existing straightforward way to test the structure of a pipeline there, since there is no mock of the `Graph` component and functions to compare with that. Let me know if the current testing is acceptable and maybe we can improve this in the future.

zijianjoy commented on frontend/src/lib/StaticGraphParser.ts:
```diff
@@ -263,3 +264,60 @@ export function createGraph(workflow: Workflow): dagre.graphlib.Graph {
+ */
+export function transitiveReduction(graph: dagre.graphlib.Graph) {
+  // safeguard against too big graphs
+  if (graph.edgeCount() > 1000 || graph.nodeCount() > 1000) {
```
Can we check if graph is null here?

Bobgy replied:
For clarification, we enabled strict nullity in typescript config, so when type does not include undefined/null, it cannot be.

zijianjoy replied:
Thank you Yuan for clarification!

zijianjoy commented on frontend/src/pages/PipelineDetails.tsx:
```diff
@@ -369,8 +388,14 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
         pathname: `/pipelines/details/${this.state.pipeline.id}/version/${versionId}`,
       });
+      const graph = await this._createGraph(selectedVersionPipelineTemplate);
+      let reducedGraph = graph ? transitiveReduction(graph) : undefined;
```
Can we use transitiveReduction(graph) as a way to return undefined if graph doesn't exist? Then it will look like 

```
let reducedGraph = transitiveReduction(graph);
```

Bobgy replied:
For discussion: personally I find current code to be more readable. If we move undefined handling inside the function, it's not clear when graph will be undefined on both places. It's not clear here, because it's a simple function call. It's also not clear in the function, because it can have many call sites.

StefanoFioravanzo replied:
@zijianjoy We could, but `graph` can be `null`, not `undefined`. This is the signature of `_createGraph`: `private async _createGraph(templateString: string): Promise<dagre.graphlib.Graph | null>`.

StefanoFioravanzo replied:
We would have to change signature of `transitiveReduction` to `transitiveReduction(graph: dagre.graphlib.Graph | null): dagre.graphlib.Graph | undefined`, which isn't that clean IMHO

StefanoFioravanzo replied:
(@Bobgy sorry my page hadn't refreshed so I wrote my comments without seeing yours first. But we are agreeing :) )

zijianjoy replied:
Thank you both for clarification! That makes sense in terms of restricting the input parameter to be not undefined.

zijianjoy commented on frontend/src/lib/StaticGraphParser.test.ts:
```diff
@@ -780,4 +787,56 @@ describe('StaticGraphParser', () => {
+      const reduced = transitiveReduction(g);
+      expect(g.nodeCount()).toEqual(3);
+      expect(g.edgeCount()).toEqual(3);
+      expect(reduced).not.toBeUndefined();
```
Simplified this to `toBeDefined()`. Reference: https://jestjs.io/docs/en/expect#tobedefined

zijianjoy commented on frontend/src/lib/StaticGraphParser.test.ts:
```diff
@@ -780,4 +787,56 @@ describe('StaticGraphParser', () => {
+    it('compares graphs edges', () => {
+      const g = createGraph(newRedundantWorkflow());
+      const reduced = transitiveReduction(graphlib.json.read(graphlib.json.write(g)));
+      expect(reduced).not.toBeUndefined();
```
Simplified this to `toBeDefined()`. Reference: https://jestjs.io/docs/en/expect#tobedefined

zijianjoy commented on frontend/src/lib/StaticGraphParser.test.ts:
```diff
@@ -780,4 +787,56 @@ describe('StaticGraphParser', () => {
+      const g = createGraph(newRedundantWorkflow());
+      const reduced = transitiveReduction(graphlib.json.read(graphlib.json.write(g)));
+      expect(reduced).not.toBeUndefined();
+      expect(compareGraphEdges(g, reduced)).not.toBeTruthy();
```
Simplified to `toBeFalsey()` https://jestjs.io/docs/en/expect#tobefalsy

Bobgy commented:
Yes, I think adding one test is enough for coverage.

Bobgy commented on frontend/src/lib/StaticGraphParser.ts:
```diff
@@ -263,3 +264,60 @@ export function createGraph(workflow: Workflow): dagre.graphlib.Graph {
+ *
+ * @param graph The dagre graph object
+ */
+export function transitiveReduction(graph: dagre.graphlib.Graph) {
```
Right now, we do not enforce explicit return value type when it can be auto inferred.

Can you elaborate rationale of your preference? We can consider that.

zijianjoy replied:
Defining a return type allows IDE to perform `type check` and make right suggestion to usage. For example: IDE will suggest to add `?` if the variable can be `undefined`. And the other method `createGraph` in the same file does specify the return type.

I think having a return type specified is a way to leverage the power of typescript (when considering typescript vs javascript). But since it is not the current practice, we don't need to enforce them in this PR, we can discuss in other channel.

Bobgy replied:
@zijianjoy IDE no longer requires return type for type checking, the typescript language server has all the information available at compile time, so inferred return type also works.

Bobgy replied:
https://www.typescriptlang.org/docs/handbook/type-inference.html
In some rare cases, inferred type may not be what we want -- that's when we "must" specify a return type.

Bobgy replied:
but it's anyway a valid preference, if you think explicit return type makes code review easier

zijianjoy replied:
Thank you for the link to type inference! I understand the situation to specify return type when inference is not what we want now. Sounds good on keeping this practice.

Bobgy commented on frontend/src/pages/RunDetails.test.tsx:
```diff
@@ -1588,6 +1588,49 @@ describe('RunDetails', () => {
+        Edge node2 to node3
+      </pre>
+    `;
+      expect(tree.getByTestId('graph')).toMatchInlineSnapshot(snap);
```
When we use snapshot testing, jest will auto generate the screenshot in place. So it's not good changing that to a variable.

Two better alternatives, use toEqual instead.
Or use snapshot-diff for the diff part, you should be able to find examples in the code base.

StefanoFioravanzo replied:
@Bobgy I am not sure I follow your argument. I mostly followed the same approach as here: https://github.com/kubeflow/pipelines/blob/cc7cac43b92fdcf6f7672978a9cd1f649fdd815c/frontend/src/pages/RunDetails.test.tsx#L645-L654

Bobgy replied:
Sure, let me explain, the string between matchinlinesnapshot is generated by jest. You moved the text to be a variable, that can prevent future regeneration of the inline snapshot, so that's not preferred.

https://jestjs.io/docs/en/snapshot-testing

StefanoFioravanzo replied:
Oh wow! This is super cool, I hadn't realized inline snapshots worked like this. Thanks!

Bobgy commented:
/lgtm
/approve

zijianjoy commented:
Thank you Stefano for this change! It helps improve pipeline visualization a lot! 👍
